### PR TITLE
Fixed bug where address required fields were required for any address type (customer, brand...)

### PR DESCRIPTION
**Pull request by LittleBigDev**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fixed bug where address required fields were required for any address type (customer, brand...). Now, multiple address types are used to read/write required fields.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-2027
| How to test?  | Go to "customers" -> "addresses" and add VAT number field as required. Then go to "catalog" -> "manufacturers and suppliers" and try to edit and save a manufacturer. You SHOULD NOT have a "required field VAT number" message.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/7986)

### CODE REVIEW

**Comment by marionf** [comment]
Hello @LittleBigDev 

It doesn't seem to work for suppliers

suppliers address

**Comment by LittleBigDev** [comment]
Can you please provide steps to reproduce this ? I tried saving a supplier and it works as well as with brands. Maybe I'm missing something.

**Comment by LittleBigDev** [comment]
@marionf : I had the same issue. I re-installed the app + composer install. Now everything is ok. Can you try the same please ?

**Comment by marionf** [comment]
@LittleBigDev, it's ok after re-installing

File: classes/Address.php [hunk 1]
```diff
@@ -520,8 +520,9 @@ public static function aliasExist($alias, $id_address, $id_customer)
     public function getFieldsRequiredDB()
     {
         $this->cacheFieldsRequiredDatabase(false);
-        if (isset(self::$fieldsRequiredDatabase['Address'])) {
-            return self::$fieldsRequiredDatabase['Address'];
+        $objectName = $this->getObjectName();
+        if (isset(self::$fieldsRequiredDatabase[$objectName])) {
+            return self::$fieldsRequiredDatabase[$objectName];
```
Comment (toutantic) [inline classes/Address.php]: This function only use methods and attributes from the parent class => should move it to ObjectModel.
+ code duplication in ObjectModel

File: classes/ObjectModel.php [hunk 2]
```diff
@@ -1030,7 +1041,10 @@ public function validateField($field, $value, $id_lang = null, $skip = array(),
 
 
         // Check if field is required
-        $required_fields = (isset(self::$fieldsRequiredDatabase[get_class($this)])) ? self::$fieldsRequiredDatabase[get_class($this)] : array();
+        $objectName = $this->getObjectName();
+        $required_fields = (isset(self::$fieldsRequiredDatabase[$objectName]))
+            ? self::$fieldsRequiredDatabase[$objectName]
+            : array();
```
Comment (toutantic) [inline classes/ObjectModel.php]: code duplication

File: classes/ObjectModel.php [hunk 3]
```diff
@@ -1157,7 +1171,10 @@ public function validateController($htmlentities = true)
     {
         $this->cacheFieldsRequiredDatabase();
         $errors = array();
-        $required_fields_database = (isset(self::$fieldsRequiredDatabase[get_class($this)])) ? self::$fieldsRequiredDatabase[get_class($this)] : array();
+        $objectName = $this->getObjectName();
+        $required_fields_database = (isset(self::$fieldsRequiredDatabase[$objectName]))
+            ? self::$fieldsRequiredDatabase[$objectName]
+            : array();
```
Comment (toutantic) [inline classes/ObjectModel.php]: code duplication

File: classes/ObjectModel.php [hunk 4]
```diff
@@ -1263,7 +1280,10 @@ public function getWebserviceParameters($ws_params_attribute_name = null)
 
         $resource_parameters = array_merge_recursive($default_resource_parameters, $this->{$ws_params_attribute_name});
 
-        $required_fields = (isset(self::$fieldsRequiredDatabase[get_class($this)]) ? self::$fieldsRequiredDatabase[get_class($this)] : array());
+        $objectName = $this->getObjectName();
+        $required_fields = (isset(self::$fieldsRequiredDatabase[$objectName])
+            ? self::$fieldsRequiredDatabase[$objectName]
+            : array());
```
Comment (toutantic) [inline classes/ObjectModel.php]: code duplication

File: classes/ObjectModel.php [hunk 5]
```diff
@@ -1369,7 +1389,10 @@ public function validateFieldsRequiredDatabase($htmlentities = true)
     {
         $this->cacheFieldsRequiredDatabase();
         $errors = array();
-        $required_fields = (isset(self::$fieldsRequiredDatabase[get_class($this)])) ? self::$fieldsRequiredDatabase[get_class($this)] : array();
+        $objectName = $this->getObjectName();
+        $required_fields = (isset(self::$fieldsRequiredDatabase[$objectName]))
+            ? self::$fieldsRequiredDatabase[$objectName]
+            : array();
```
Comment (toutantic) [inline classes/ObjectModel.php]: code duplication

File: classes/Address.php (hunk 1 above)
Comment (LittleBigDev) [reply classes/Address.php]: `ObjectModel::getFieldsRequiredDB()` already exists and has a different behavior : it looks for fields in DB (`ps_required_field` table).
`Address::getFieldsRequiredDB()` tho, looks for "cached" (saved in a protected static property) required fields only. It never looks into DB.
Required fields seem to be cached no matter what. So this is functionally ok. But maybe we should clean this up... What do you think ?

File: classes/ObjectModel.php [hunk 6]
```diff
@@ -1460,12 +1485,18 @@ public function addFieldsRequiredDatabase($fields)
             return false;
         }
 
-        if (!Db::getInstance()->execute('DELETE FROM '._DB_PREFIX_.'required_field WHERE object_name = \''.get_class($this).'\'')) {
+        $objectName = $this->getObjectName();
+        if (!Db::getInstance()->execute(
+            'DELETE FROM ' . _DB_PREFIX_ . "required_field WHERE object_name = '$objectName'")
```
Comment (eternoendless) [inline classes/ObjectModel.php]: `$objectName` MUST be escaped. I think you can use `DB::getInstance()->escape()` for that.

File: classes/ManufacturerAddress.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2017 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+/**
+ * Class ManufacturerAddressCore
```
Comment (eternoendless) [inline classes/ManufacturerAddress.php:28]: Please write a more descriptive PHPDoc header 🙂

**Review by eternoendless — CHANGES_REQUESTED** [review]

File: classes/ObjectModel.php (hunk 6 above)
Comment (LittleBigDev) [reply classes/ObjectModel.php]: Ok, done.

File: classes/ManufacturerAddress.php (hunk 7 above)
Comment (LittleBigDev) [reply classes/ManufacturerAddress.php:28]: Done.

**Review by eternoendless — DISMISSED** [review]

File: classes/ObjectModel.php [hunk 8]
```diff
@@ -1468,7 +1468,13 @@ public function getCachedFieldsRequiredDatabase($all = false)
     {
         $this->cacheFieldsRequiredDatabase($all);
 
-        return $all ? self::$fieldsRequiredDatabase : self::$fieldsRequiredDatabase[$this->getObjectName()];
+        if ((bool)$all) {
```
Comment (eternoendless) [inline classes/ObjectModel.php]: The explicit casting to `(bool)` is not needed IMHO.

File: classes/ObjectModel.php [hunk 9]
```diff
@@ -1468,7 +1468,13 @@ public function getCachedFieldsRequiredDatabase($all = false)
...
+            return self::$fieldsRequiredDatabase;
+        }
+
+        return !empty(self::$fieldsRequiredDatabase[$this->getObjectName()])
+            ? self::$fieldsRequiredDatabase[$this->getObjectName()]
+            : array();
```
Comment (eternoendless) [inline classes/ObjectModel.php]: What about storing `self::$fieldsRequiredDatabase[$this->getObjectName()` in a variable and then checking `!empty()` on it?
That would prevent you from calling `$this->getObjectName()` twice.

**Review by eternoendless — DISMISSED** [review]

File: classes/ObjectModel.php (hunk 8 above)
Comment (LittleBigDev) [reply classes/ObjectModel.php]: Ok done.

File: classes/ObjectModel.php (hunk 9 above)
Comment (LittleBigDev) [reply classes/ObjectModel.php]: I'd have to test `!empty()` before storing the var, and one more time to determine the return. 
I created `$objectName` instead

File: classes/ObjectModel.php [hunk 10]
```diff
@@ -1447,6 +1458,28 @@ public function cacheFieldsRequiredDatabase($all = true)
...
     /**
+     * Get required fields list for this model or for all the models
+     *
+     * @param bool $all : whether it should return required fields for this model or all the models
+     *
+     * @return array
+     */
+    public function getCachedFieldsRequiredDatabase($all = false)
+    {
+        $this->cacheFieldsRequiredDatabase($all);
```
Comment (eternoendless) [inline classes/ObjectModel.php:1468]: Why is it called here?

Comment (LittleBigDev) [reply classes/ObjectModel.php:1468]: Before, it was called in `AddressCore::getFieldsRequiredDB()`. But with refactoring, the call was moved in parent class.

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by vincentbz** [comment]
Hi,

Installation failed with that PR : 

image

**Comment by marionf** [comment]
J'ai le même problème que @vincentbz de mon côté

**Comment by LittleBigDev** [comment]
Cannot reproduce. I asked the steps to reproduce in Jira.

**Comment by marionf** [comment]
The installation is ok after re-installing composer.

I have an SQL error when I try to add a required field in the customer address
pr7986

**Comment by marionf** [comment]
It's ok for SQL error.
I check vat_number as required fields in the customer address
I try to add an address in BO, I don't have the star for the vat_number field but if I don't fill it I can't save and the error ask me to fill the vat_nimber field.
I try to add an address in FO, the vat_number field is marked optionnal and if I don't fill it I can save without error message.

**Comment by LittleBigDev** [comment]
Should be fixed. But code is a bit spaghetti. Meticulous tests are needed to ensure no side effect was introduced.

**Comment by marionf** [comment]
I will test after the code review ;)

**Review by tomlev — DISMISSED** [review]

**Comment by marionf** [comment]
On develop, If I add required fields and I try to proceed to checkout with an address without these fields, I have error 500 for each field.
capture du 2017-10-11 14-47-12

On this PR, if I try to proceed to checkout with an address without these fields, I have no error and I can complete the order.

**Comment by LittleBigDev** [comment]
Rebased with (and adapted to) frontend fixes. New review required.

**Review by tomlev — APPROVED** [review]

**Comment by LittleBigDev** [comment]
Reviewed 1 of 9 files at r1, 4 of 5 files at r3, 2 of 4 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7986)*

**Comment by tomlev** [comment]
Reviewed 1 of 9 files at r1, 4 of 5 files at r3, 2 of 4 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7986)*

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by eternoendless** [comment]
Thank you @LittleBigDev

**Pull request merged**